node/router/src/inbound.rs

```diff
@@ -142,7 +145,10 @@ pub trait Inbound<N: Network>: Reading + Outbound<N> {
                 bail!("Peer '{peer_ip}' is not following the protocol")
             }
             Message::Disconnect(message) => {
-                bail!("{:?}", message.reason)
+                // The peer informs us that they had disconnected. Disconnect from them too.
+                debug!("Peer '{peer_ip}' decided to disconnect due to '{:?}'", message.reason);
+                self.router().disconnect(peer_ip);
+                Ok(())
```

Comment:
@ljedrz @raychu86 can you double check this? `Ok(())` maintains the connection

Reply:
This new logic skips the call to 

`Outbound::send(self, peer_ip, Message::Disconnect(DisconnectReason::ProtocolViolation.into()));`. I believe this is fine because in theory the peer has already disconnected you since they are sending a Disconnect message.

Reply:
I agree it should work but like [the other change](https://github.com/ProvableHQ/snarkOS/pull/3540#discussion_r2105443846) in `inbound` it complicates the interface.  In this case the `self.router().disconnect(peer_ip)` has to get copied from `process_message{_internal}` to `inbound`.  This is the only call to `disconnect()` in `inbound`.